linen_examples/pixelcnn/train.py

```diff
@@ -0,0 +1,299 @@
+  loss, grad = grad_fn(optimizer.target)
+  grad = lax.pmean(grad, 'batch')
+  optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
+
+  # Compute exponential moving average (aka Polyak decay)
+  ema_decay = FLAGS.polyak_decay
+  ema = jax.tree_multimap(lambda ema, p: ema * ema_decay + (1 - ema_decay) * p,
+                          ema, optimizer.target)
+
+  metrics = {'loss': lax.pmean(loss, 'batch'), 'learning_rate': lr}
+  return optimizer, ema, metrics
+
+
+def eval_step(params, batch):
+  images = batch['image']
+  logits = model(dropout_p=0.).apply({'param': params}, images)
+  return {'loss': lax.pmean(neg_log_likelihood_loss(logits, images), 'batch')}
+
+
+def load_and_shard_tf_batch(xs):
+  local_device_count = jax.local_device_count()
+
+  def _prepare(x):
+    # Use _numpy() for zero-copy conversion between TF and NumPy.
+    x = x._numpy()  # pylint: disable=protected-access
+    return x.reshape((local_device_count, -1) + x.shape[1:])
+
+  return jax.tree_map(_prepare, xs)
+
+
+def restore_checkpoint(optimizer, ema):
+  return checkpoints.restore_checkpoint(FLAGS.model_dir, (optimizer, ema))
+
+
+def save_checkpoint(optimizer, ema, step):
+  optimizer, ema = jax_utils.unreplicate((optimizer, ema))
+  checkpoints.save_checkpoint(FLAGS.model_dir, (optimizer, ema), step, keep=3)
+
+
+def train():
+  """Train model."""
+  batch_size = FLAGS.batch_size
+  n_devices = jax.device_count()
+  if jax.host_count() > 1:
+    raise ValueError('PixelCNN++ example should not be run on more than 1 host'
+                     ' (for now)')
+  if batch_size % n_devices > 0:
+    raise ValueError('Batch size must be divisible by the number of devices')
+
+  train_summary_writer, eval_summary_writer = get_summary_writers()
+
+  # Load dataset
+  data_source = input_pipeline.DataSource(
+      train_batch_size=batch_size, eval_batch_size=batch_size)
+  train_ds = data_source.train_ds
+  eval_ds = data_source.eval_ds
+
+  # Create dataset batch iterators
+  train_iter = iter(train_ds)
+  eval_iter = iter(eval_ds)
+
+  # Compute steps per epoch and nb of eval steps
+  steps_per_epoch = data_source.TRAIN_IMAGES // batch_size
+  steps_per_eval = data_source.EVAL_IMAGES // batch_size
+  steps_per_checkpoint = steps_per_epoch * 10
+  num_steps = steps_per_epoch * FLAGS.num_epochs
+
+  # Create the model using data-dependent initialization. Don't shard the init
+  # batch.
+  assert FLAGS.init_batch_size <= batch_size
+  init_batch = next(train_iter)['image']._numpy()[:FLAGS.init_batch_size]
+
+  rng = random.PRNGKey(FLAGS.rng)
+  rng, init_rng = random.split(rng)
+  rng, dropout_rng = random.split(rng)
+
+  initial_variables = model().init({
+      'param': init_rng,
+      'dropout': dropout_rng
+  }, init_batch)['param']
+  optimizer_def = optim.Adam(
+      learning_rate=FLAGS.learning_rate, beta1=0.95, beta2=0.9995)
+  optimizer = optimizer_def.create(initial_variables)
+
+  # TODO(marcvanzee): Restoring checkpoints is broken. Fix this.
```

Comment:
Yes I agree fixing it would be good. But since the example works fine for now, it does not seem blocking for this PR. I'll looking into it, and once it is fixed I will uncomment this. WDYT?

Reply:
Yeah, agreed, if the problem is unrelated to PixelCNN I don't think it's worth blocking on. I thought @levskaya made it sound as though it wouldn't be too difficult to fix though :).